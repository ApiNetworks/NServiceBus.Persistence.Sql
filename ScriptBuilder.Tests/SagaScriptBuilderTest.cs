using System.Collections.Generic;
using System.IO;
using System.Text;
using ApprovalTests;
using NServiceBus.SqlPersistence;
using NUnit.Framework;

[TestFixture]
public class SagaScriptBuilderTest
{
    [Test]
    public void BuildCreateScript()
    {
        var sagas = new List<SagaDefinition>
        {
            new SagaDefinition
            {
                Name = "theSaga",
                MappedProperties = new List<string>
                {
                    "Property1",
                    "Property2"
                }
            }
        };

        var builder = new StringBuilder();
        using (var writer = new StringWriter(builder))
        {
            SagaScriptBuilder.BuildCreateScript("theschema", "theendpointname", sagas, s => writer);
        }
        var script = builder.ToString();

        SqlValidator.Validate(script);
        Approvals.Verify(script);
    }

    [Test]
    public void BuildDropScript()
    {
        var sagas = new List<string>
        {
            "theSaga"
        };

        var builder = new StringBuilder();
        using (var writer = new StringWriter(builder))
        {
            SagaScriptBuilder.BuildDropScript("theschema", "theendpointname", sagas, s => writer);
        }
        var script = builder.ToString();
        SqlValidator.Validate(script);
        Approvals.Verify(script);
    }
}
