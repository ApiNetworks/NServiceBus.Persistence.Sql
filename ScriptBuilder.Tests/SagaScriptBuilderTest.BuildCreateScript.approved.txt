
declare @tableName nvarchar(max) = '[' + @schema + '].[' + @endpointName + '.theSaga]';


IF NOT EXISTS
(
    SELECT *
    FROM sys.objects
    WHERE
        object_id = OBJECT_ID(@tableName) AND
        type in (N'U')
)
BEGIN
DECLARE @createTable nvarchar(max);
SET @createTable = N'
    CREATE TABLE ' + @tableName + N'(
	    [Id] [uniqueidentifier] NOT NULL PRIMARY KEY,
	    [Originator] [nvarchar](255) NULL,
	    [OriginalMessageId] [nvarchar](255) NULL,
	    [Data] [xml] NOT NULL,
	    [PersistenceVersion] [nvarchar](23) NOT NULL,
	    [SagaTypeVersion] [nvarchar](23) NOT NULL
    )
';
exec(@createTable);
END

declare @dropIndexQuery nvarchar(max);
select @dropIndexQuery =
(
    SELECT 'DROP INDEX ' + ix.name + ' ON ' + @tableName + '; '
    FROM sysindexes ix
    WHERE
		ix.Id = (select object_id from sys.objects where name = @tableName) AND
	    ix.Name IS NOT null AND
	    ix.Name LIKE 'PropertyIndex_%' AND
	    ix.Name <> 'PropertyIndex_Property1'
    for xml path('')
);
exec sp_executesql @dropIndexQuery

IF NOT EXISTS
(
    SELECT *
    FROM sys.indexes
    WHERE
        name = 'PropertyIndex_Property1' AND
        object_id = OBJECT_ID(@tableName)
)
BEGIN
DECLARE @createIndex nvarchar(max);
SET @createIndex = N'
    CREATE  SELECTIVE XML INDEX PropertyIndex_Property1
    ON ' + @tableName + '(Data)
    FOR
    (
        Property1 = ''/Data/Property1'' AS XQUERY ''xs:string'' SINGLETON
    )
';
exec(@createIndex);
END
